### PR TITLE
Storing Courses in DB

### LINKED ISSUE #106: Figure out way to reduce database row count to <= 10,000
* Heroku Postgres has a limit of 10k rows for the free tier and currently our app only uses one endpoint so maybe we could reduce the data inserted into the DB

### PR DESCRIPTION
**Pull request by resurl**
# Changes
- replaced CourseSection table with Course table in `setup.ts`, following the structure of the json
- translated `CourseSection` controller functions to `Course` controller functions
- refactored scraper to get rid of duplicate code 
- bug fix in scraper where coreqText wasn't being stored 
- ran scraper again to update `output.json`
- added api endpoints for `Course`

## Motivation and Context
to reduce the # of rows we store in heroku

## How Has This Been Tested?
- has only been tested locally (docker is giving my computer issues atm 😞)
- ran `npm run start:localwindows` in `server/` and checked the database in pgadmin
- (for API) compared the result of querying `coursesection` endpoints and `course` endpoints

## Screenshots (if appropriate):
<details>
<summary>Storing in database</summary>

image
</details>

<details>
<summary>Querying /api/sections/ (using new course controller functions)</summary>

image
</details>

<details>
<summary>Querying /api/section/CPSC/210</summary>

image
</details>

<details>
<summary>Querying /api/section/1/CPSC/210</summary>

image
</details>

## Todo
- implement getCourseSection in `server/src/controller/course.ts` ✅
- add the express routes ✅
- test the API ✅

# Related Issues
- Closes #106

### CODE REVIEW

**Comment by resurl** [comment]
(also i remember john had some trouble accessing my changes last time--i forked the repo so i think you'll need to do `gh pr checkout 126` in the github CLI to look at it directly, sorry for the inconvenience lol)

**Comment by johnagl** [comment]
Nice, this looks like it's progressing well, some initial thoughts I have:

1) Let's try to keep the schema of a Course Section the same as before (if possible) so that we don't have to make too many changes to the scheduler algorithm

2) We will have to change the API requests on the frontend to the new endpoints you're adding (I can help out with this when the endpoints are ready to be used) 

3) Make sure to keep this branch up to date so you don't run into merge conflicts later

**Comment by resurl** [comment]
thanks for looking over it @johnagl ! i finished the endpoints for course so it should give the same output as the coursesection endpoints, except the api names are a bit different 
- `/sections` does the same thing, gets all sections
- `/section/:coursedept/:coursenumber` ➡ `/course/:coursedept/:coursenumber`
- `/section/:term/:coursedept/:coursenumber` ➡ `/course/:term/:coursedept/:coursenumber`

or i could just make the course endpoints the same as the coursesection endpoints?

**Comment by johnagl** [comment]
> thanks for looking over it @johnagl ! i finished the endpoints for course so it should give the same output as the coursesection endpoints, except the api names are a bit different
> 
> * `/sections` does the same thing, gets all sections
> * `/section/:coursedept/:coursenumber` ➡ `/course/:coursedept/:coursenumber`
> * `/section/:term/:coursedept/:coursenumber` ➡ `/course/:term/:coursedept/:coursenumber`
> 
> or i could just make the course endpoints the same as the coursesection endpoints?

Honestly, if we can keep the same endpoints then it means we don't have to tweak the frontend API requests. As long as the new API's are returning the same information then it should be fine to just keep the same ones.

**Comment by johnagl** [comment]
> > thanks for looking over it @johnagl ! i finished the endpoints for course so it should give the same output as the coursesection endpoints, except the api names are a bit different
> > 
> > * `/sections` does the same thing, gets all sections
> > * `/section/:coursedept/:coursenumber` ➡ `/course/:coursedept/:coursenumber`
> > * `/section/:term/:coursedept/:coursenumber` ➡ `/course/:term/:coursedept/:coursenumber`
> > 
> > or i could just make the course endpoints the same as the coursesection endpoints?
> 
> Honestly, if we can keep the same endpoints then it means we don't have to tweak the frontend API requests. As long as the new API's are returning the same information then it should be fine to just keep the same ones.

Keeping the same endpoints might not be that easy though - so I would say go with the solution that is most doable and let me know if we need to change the frontend api requests.

**Comment by resurl** [comment]
@johnagl i think we should be able to keep the api requests the same in the frontend! the controllers for course right now should return the same information as the coursesections ones now

will it be okay to delete the coursesection code in this pr?

**Comment by johnagl** [comment]
> @johnagl i think we should be able to keep the api requests the same in the frontend! the controllers for course right now should return the same information as the coursesections ones now
> 
> will it be okay to delete the coursesection code in this pr?

Yup that should be fine!

File: server/src/database/setup.ts
Comment (johnagl) [inline server/src/database/setup.ts:24]: Okay, correct me if I'm wrong, but it looks like we aren't deleting the CourseSection table yet - took a look at what's in my local db. I think it's because you've replaced `CourseSection` with `Course`. We should probably keep `CourseSection` here initially.

**Review by johnagl — APPROVED** [review]
Took a look at the code, and tested, looks good - besides the non deletion of the CourseSection table which I commented on above. 

@jacksonyuanjx Since you have experience with the database and first pointed out the 10k row limit - do you know if we are able to start inserting rows again once there are less than 10k rows? 

I'll approve this for now @resurreccionl,  lets wait for @jacksonyuanjx response before merging just to make sure we can actually insert stuff into the DB.

**Comment by johnagl** [comment]
Also I had to follow this https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/checking-out-pull-requests-locally @resurreccionl to checkout the PR. I'm not sure why your branches are setup like that ...

**Comment by jacksonyuanjx** [comment]
> Took a look at the code, and tested, looks good - besides the non deletion of the CourseSection table which I commented on above.
> 
> @jacksonyuanjx Since you have experience with the database and first pointed out the 10k row limit - do you know if we are able to start inserting rows again once there are less than 10k rows?
> 
> I'll approve this for now @resurreccionl, lets wait for @jacksonyuanjx response before merging just to make sure we can actually insert stuff into the DB.

Yep, I read somewhere that if we fix the 10k rows issue, then we can insert again.

File: server/src/database/setup.ts
Comment (resurl) [reply server/src/database/setup.ts:24]: nope that makes sense! added the coursesection

**Pull request merged**
